include/boost/gil/algorithm.hpp

```diff
@@ -143,30 +143,24 @@ namespace std {
 template<typename T, typename CS>
 BOOST_FORCEINLINE
 auto copy(
-    boost::gil::pixel<T, CS>* first,
-    boost::gil::pixel<T, CS>* last,
+    boost::gil::pixel<T, CS> const* first,
+    boost::gil::pixel<T, CS> const* last,
     boost::gil::pixel<T, CS>* dst)
     ->  boost::gil::pixel<T, CS>*
 {
-    auto p = std::copy((unsigned char*)first, (unsigned char*)last, (unsigned char*)dst);
+    auto p = std::copy(
+        reinterpret_cast<unsigned char const*>(first),
+        reinterpret_cast<unsigned char const*>(last),
+        reinterpret_cast<unsigned char*>(dst));
     return reinterpret_cast<boost::gil::pixel<T, CS>*>(p);
 }
 
-/// \ingroup STLOptimizations
-/// \brief Copy when both src and dst are interleaved and of the same type can be just memmove
-template<typename T, typename CS>
-BOOST_FORCEINLINE
-auto copy(const boost::gil::pixel<T,CS>* first, const boost::gil::pixel<T,CS>* last,
-     boost::gil::pixel<T,CS>* dst) -> boost::gil::pixel<T,CS>*
-{
-    return (boost::gil::pixel<T,CS>*)std::copy((unsigned char*)first,(unsigned char*)last, (unsigned char*)dst);
-}
 } // namespace std
 
 namespace boost { namespace gil {
 namespace detail {
 template <typename I, typename O> struct copy_fn {
-    BOOST_FORCEINLINE I operator()(I first, I last, O dst) const { return std::copy(first,last,dst); }
+    BOOST_FORCEINLINE O operator()(I first, I last, O dst) const { return std::copy(first,last,dst); }
```

Comment:
Hmm, amazing it remained unrevealed for long time.
How did you find it out?

Reply:
> Hmm, amazing it remained unrevealed for long time.

I was also surprised after realizing this bug has been there for at least 15 years. I guess the reason is copy_fn is not used in many places and as long as the output iterator is implicitly convertible to the input iterator (as it is the case for pointers which only differ in their const-qualifier), everything is fine.

> How did you find it out?

Just me, the algorithm header and a cup of coffee.